pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+	if claim.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating updateClaimPhaseWithEvent[%s]: phase %s already set", claimToClaimKey(claim), phase)
+		return claim, nil
+	}
+
+	// Emit the event only when the status change happens, not everytime
+	// syncClaim is called.
+	glog.V(3).Infof("Claim %q changed status to %q: %s", claimToClaimKey(claim), phase, message)
+	ctrl.eventRecorder.Event(claim, eventtype, reason, message)
+
+	newClaim, err := ctrl.updateClaimPhase(claim, phase)
+	return newClaim, err
+}
+
+func (ctrl *PersistentVolumeController) updateVolumePhase(volume *api.PersistentVolume, phase api.PersistentVolumePhase) (*api.PersistentVolume, error) {
+	glog.V(5).Infof("Updating PersistentVolume[%s]: set phase %s", volume.Name, phase)
+	if volume.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating PersistentVolume[%s]: phase %s already set", volume.Name, phase)
+		return volume, nil
+	}
+
+	clone, err := conversion.NewCloner().DeepCopy(volume)
+	if err != nil {
+		return nil, fmt.Errorf("Error cloning claim: %v", err)
+	}
+	volumeClone, ok := clone.(*api.PersistentVolume)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected volume cast error : %v", volumeClone)
+	}
+
+	volumeClone.Status.Phase = phase
+	newVol, err := ctrl.kubeClient.Core().PersistentVolumes().UpdateStatus(volumeClone)
+	if err != nil {
+		glog.V(5).Infof("Updating PersistentVolume[%s]: set phase %s failed: %v", volume.Name, phase, err)
+		return newVol, err
+	}
+	glog.V(5).Infof("Updating PersistentVolume[%s]: phase %s set", volume.Name, phase)
+	return newVol, err
+}
+
+func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *api.PersistentVolume, phase api.PersistentVolumePhase, eventtype, reason, message string) (*api.PersistentVolume, error) {
+	glog.V(5).Infof("Updating updateVolumePhaseWithEvent[%s]: set phase %s", volume.Name, phase)
+	if volume.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating updateVolumePhaseWithEvent[%s]: phase %s already set", volume.Name, phase)
+		return volume, nil
+	}
+
+	// Emit the event only when the status change happens, not everytime
+	// syncClaim is called.
+	glog.V(3).Infof("Volume %q changed status to %q: %s", volume.Name, phase, message)
+	ctrl.eventRecorder.Event(volume, eventtype, reason, message)
+
+	newVol, err := ctrl.updateVolumePhase(volume, phase)
+	return newVol, err
+}
+
+func (ctrl *PersistentVolumeController) bindVolumeToClaim(volume *api.PersistentVolume, claim *api.PersistentVolumeClaim) (*api.PersistentVolume, error) {
+	glog.V(5).Infof("Updating PersistentVolume[%s]: binding to %q", volume.Name, claimToClaimKey(claim))
+
+	// Check if the volume was already bound
+	boundByController := false
+	if !isVolumeBoundToClaim(volume, claim) {
+		boundByController = true
+	}
+
+	// Save the PV only when any modification is neccesary.
+	if volume.Spec.ClaimRef == nil || volume.Spec.ClaimRef.Namespace != claim.Namespace ||
```

Comment:
it might save some brain cells to make bools for these predicates like `nilClaimRef` and `nameSpaceMismatch`